yaze.1
.\"	$Id: yaze.1,v 1.2 2004/01/11 16:11:17 fdc Exp $	
.TH yaze 1 "10 October 1995" "Frank\'s Hacks" "Local commands"
.UC 4
.SH NAME
yaze \- yet another z80 emulator
.SH SYNOPSIS
.PU
.ll +8
.B yaze
.RB [ -v ]
.RB [ -b\fIbootfile ]
.RB [ -l\fIloadadr ]
.RB [ -p\fIbasepage ]
.RB [ -s\fIstartup ]
.RB [ -z\fIZ3ENV ]
.I command...
.ll -8
.br
.SH DESCRIPTION
.B yaze
is designed to provide an exact simulation of the Z80 microprocessor
on a Unix system.  In addition to the instruction-set emulator, a
CP/M basic i/o system is provided that can reference Unix directories
and Unix files containing images of CP/M disks.  When a genuine CP/M
operating system (ccp + bdos) or a clone system is loaded into the
emulator, a complete CP/M system is available, running on a Unix host.

The complete documentation is contained in the file yaze.doc.
.SH OPTIONS
.TP
.B -v
causes a summary of the system configuration to be displayed after
loading.
.TP
.B -b
gives the name of a file to be loaded into the emulated processor's
ram before the emulation is started.  The file is loaded at the
address given by the \fB-l\fR option, if one is present, or else at
the basepage (see \fB-p\fR).  If a bootfile is not specified with the
\fB-b\fR option, yaze looks for the file yaze.boot in the current
directory first, then in /usr/local/lib.
.TP
.B -l
the hexadecimal address at which to load and start the bootstrap file,
if a separate bootstrap is necessary.
.TP
.B -p
the top 2 hex digits of the location of the CP/M console command
processor.  Also the location to which bootfile is loaded if a
\fB-l\fR option is not present.
.TP
.B -s
a file containing monitor commands that are to be executed before
starting the emulator.  If no \fB-s\fR option is present, .yazerc from
the current directory or from the user's home directory is taken.
.TP
.B -z
the hexadecimal address in the emulated processor's ram of a 1KByte
area that should be reserved for use by an extended CP/M clone such as
ZCPR3.
.SH MONITOR COMMANDS
When the emulator is started it executes commands first from the
startup file, then from the tail of the command line (any strings
which follow the options) and then from the tty if neither the startup
file nor the command line included a "go" command.  Control can be
returned to the monitor by executing the CP/M sys.com command or, if
yaze was compiled with -DDEBUG, by sending it a SIGINT signal.

Command names can be shortened to a unique abbreviation.  If
\fByaze\fR was compiled with GNU Readline, command-line editing,
filename completion and history recall are available - see the
readline documentation.

The monitor commands are described in more detail in the file
yaze.doc.
.TP
.B help
Display a command list
.TP
.B help \fIcmd\fR
Give details about \fIcmd\fR
.TP
.B attach
Attach CP/M device to a Unix file
.TP
.B detach
Detach CP/M device from file
.TP
.B mount
Mount a Unix file or directory as a CP/M disk
.TP
.B umount
Unmount a CP/M disk
.TP
.B create
Create a new disk
.TP
.B interrupt
Set user interrupt key
.TP
.B go
Start/Continue CP/M execution
.TP
.B !
Execute a Unix command
.TP
.B quit
Terminate yaze
.TP
.B time
Display elapsed time since last `time' command
.SH SEE ALSO
.B cdm(1)
.SH AUTHOR
Frank D. Cringle (fdc@cliwe.ping.de).


ChangeLog
**** Version 1.14 released ****

2004-04-23  Frank D. Cringle  <fdc@sabre.ping.de>

	* yaze.c, Makefile (VERSION): update version string.

	* simz80.pl: simulate the interrupt register (IR).  (Michael Haardt).

	* bios.c (bios): case 254 incremented pc twice in a macro argument
	(Carl Mascott).

**** Version 1.13 released ****

2004-01-24  Frank D. Cringle  <fdc@cliwe.ping.de>

	* yaze.c, Makefile (VERSION): update version string.

	* simz80.h: pagetable definition did not match the definition in
	yaze.c (Michael Haardt).

**** Version 1.12 released ****

2004-01-12  Frank D. Cringle  <fdc@cliwe.ping.de>

	* Makefile: host2cpm.tcl should be included in the tarball.

**** Version 1.11 released ****

2004-01-11  Frank D. Cringle  <fdc@cliwe.ping.de>

	* monitor.c (CLK_TCK): conditional definition.

	* cdm.c: add 'ambiguous filename -> unix directory' functionality,
	contributed by Carl Mascott.

	* bios.c bios.h monitor.c: apply patch from Guenter Radestock so
	that stdin can be a pipe.
	(readsec): fix lru sort bug reported by Carl Mascott.

	* simz80.pl (OUTP): fix register selection (Michael Haardt).

	* io.c yaze.c: fixes for CP/M 3.0 by Michael Haardt.

	* host2cpm.tcl: copies host directories to/from a cpm disk image
	file, contributed by Gary L. Howell.

	* Makefile: 'make tar' modified so that the test sources are
	converted from LF to CRLF.  Add host2cpm.tcl to DISTRIB.




**** Version 1.10 released ****

1998-01-27  Frank D. Cringle  <fdc@cliwe.ping.de>

	* simz80.pl: set the flags correctly for LD A,I/R and
	CPI/CPIR/CPD/CPDR (reported by Sean Young).

	* monitor.c (mountdir): fix fencepost error when setting cp[15],
	as reported by Richard Hirst.

1998-01-25  Frank D. Cringle  <fdc@cliwe.ping.de>

	* simz80.pl: change direct and indirect references to the RAM()
	macro, because it now evaluates its operand more than once if MMU
	is in use (contributed by Michael Haardt).

	* yaze.c, simz80.h: Make BIOS optional and add MMU initialisation
	(contributed by Michael Haardt).

Sun Apr 28 11:30:52 1996  Frank D. Cringle  <fdc@cliwe.ping.de>

	* cdm.c (getblock): rename index to idx (avoids a warning about
	shadowing index()).
	(main): correct a type in the !USE_GNU_READLINE case.

Thu Apr 25 15:32:24 1996  Frank D. Cringle  <fdc@cliwe.ping.de>

	* bios.c (constat, lststat): select() may modify timeval, so
	reinitialise it.
	(readsec,writesec): protect against undefined or unmounted disk.
	(bios): Add BgiiFix.

	* monitor.c (mount): use MAP_FILE flag in mmap() call (linux-ism?).

	* yaze.c (load_cpm): correctly calculate available memory if
	loadadr is specified.

	* simz80.pl (CALLC): rename symbol adr to avoid warning msg.

	* cdm.c (srchdir): ignore cpm-3 byte count.




yaze.c
/* yaze - yet another Z80 emulator.
   Copyright (C) 1995,1998  Frank D. Cringle.

Yaze is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "simz80.h"
#include "yaze.h"

/* 	$Id: yaze.c,v 1.5 2004/04/23 10:05:25 fdc Exp $	 */

#ifndef lint
static char vcid[] = "$Id: yaze.c,v 1.5 2004/04/23 10:05:25 fdc Exp $";
#endif /* lint */


#define VERSION	"1.14"

/* Z80 registers */
WORD af[2];			/* accumulator and flags (2 banks) */
int af_sel;			/* bank select for af */

struct ddregs regs[2];		/* bc,de,hl */
int regs_sel;			/* bank select for ddregs */

WORD ir;			/* other Z80 registers */
WORD ix;
WORD iy;
WORD sp;
WORD pc;
WORD IFF;
WORD IM;

BYTE ram[MEMSIZE*1024];		/* Z80 memory space */
#ifdef MMU
BYTE *pagetable[16];		/* MMU page table */
#endif

#ifndef LIBDIR
#define LIBDIR "/usr/local/lib/"
#endif

static char *bootfile = "yaze.boot";
static char *startup = ".yazerc";
static char *progname;
static long loadadr = -1;
#ifdef BIOS
static long basepage = 0xe4;
static int vflag;
long z3env;

WORD	ccp_base;		/* base address of ccp */
WORD	bdos_base;		/* base address of bdos */
WORD	bios_base;		/* base address of bios */
WORD	bios_top;		/* end of bios, start of global work area */

WORD	dirbuff;		/* common directory buffer for all disks */
WORD	dptable;		/* base of disk parameter table */
BYTE	*global_alv;		/* global allocation vector */
#endif

static void
usage(void)
{
    fprintf(stderr,
            "Usage: %s {flags} {commands}\n"
            "           -b <file>       boot from this file\n"
            "           -l <xxxx>       load bootfile at given (hex) address\n"
#ifdef BIOS
            "           -p <xx>         base page (top 2 hex digits of ccp base)\n"
#endif
            "           -s <file>       startup file\n"
#ifdef BIOS
            "           -v              verbose\n"
            "           -z <xxxx>       (hex) address of z3env (if desired)\n\n"
#endif
            ,progname);
    exit (1);
}

/* This routine loads the bootstrap file into the simulated processor's ram.

   If a load address is given with the -l flag, the boot file is
   loaded to that address, the z80 pc is set to it and hl is set to
   the ccp base address.  The bootfile might be a standalone program
   which ignores the bios vectors or it might contain a
   self-relocating ccp/bdos (clone), in which case it should relocate
   itself and jump to the bios cold boot vector.

   Otherwise, the boot file can contain 1 or 2 copies of cp/m.  If it
   contains 1 copy the specified or default basepage must correspond
   to the run-time location of the cp/m image.  If it contains 2
   copies of the same code assembled to run at different locations,
   the code can and will be relocated to fit the basepage in use. */

static void
load_cpm(const char *bfname)
{
    FILE *cpm;
#ifdef BIOS
    int i, skip = 0;
    WORD offs;
    struct stat	sb;
#endif

    if ((cpm = fopen(bfname, "r")) == NULL) {
	char *p = xmalloc(strlen(bfname) + strlen(LIBDIR) + 1);
	strcpy(p, LIBDIR);
	strcat(p, bfname);
	if ((cpm = fopen(p, "r")) == NULL)
	    free(p);
	else
	    bfname = p;
    }
    if (cpm == NULL) {
	perror(bfname);
	exit(1);
    }
    if (loadadr >= 0) {
	fread(ram+loadadr, 1, 64*1024-loadadr, cpm);
	fclose(cpm);
	pc = loadadr;
#ifdef BIOS
	regs[regs_sel].hl = ccp_base;
#endif
	return;
    }
#ifdef BIOS
    if (fstat(fileno(cpm), &sb) != 0) {
	perror(bfname);
	exit(1);
    }
    if ((sb.st_size >= CPM_LENGTH+0x880 && sb.st_size < 2*CPM_LENGTH) ||
	(sb.st_size >= 2*CPM_LENGTH+0x880)) {
	skip = 0x880;
	fseek(cpm, skip, SEEK_SET);
    }
    fread(ram + ccp_base, CPM_LENGTH, 1, cpm);
    if (sb.st_size < 2*CPM_LENGTH) {
	/* single, non-relocatable copy */
	if (ram[ccp_base] == 0xc3 && (ram[ccp_base+2]&0xfc) != basepage)
	    printf("Warning: %s appears to need -p %2x\n", bfname,
		   ram[ccp_base+2]&0xfc);
    }
    else {
	if ((sb.st_size & 1) || ram[ccp_base] != 0xc3) {
	    fprintf(stderr, "%s does not appear to contain 2 cp/m images\n", bfname);
	    exit(1);
	}
	fseek(cpm, skip + sb.st_size/2, SEEK_SET);
	/* this loses if the initial jump points into the 2nd half of ccp */
	offs = (ccp_base >> 8) - (ram[ccp_base + 2] & 0xfc);
	for (i = 0; i < CPM_LENGTH; i++) {
	    if ((BYTE) getc(cpm) != ram[ccp_base + i])
		ram[ccp_base + i] += offs;
	}
    }
    fclose(cpm);
    printf("Running '%s'\n", bfname);
    ram[3] = 0x95;			/* iobyte */
    ram[4] = 0;				/* disk A: */
    pc = bios_base;
#endif
}

#ifdef BIOS
#define N_VEC	17			/* number of bios entry points */

static void
setvectors(void)
{
    int i;
    long cpm_brk = 64*1024;

    if (z3env == cpm_brk - 1024)	/* z3env at top of memory? */
	cpm_brk = z3env;
    cpm_brk -= 128;			/* space for directory buffer */
    dirbuff = cpm_brk;
    cpm_brk -= 16*(16+15);		/* space for 16 disk parameter headers
					   and disk parameter blocks */
    dptable = cpm_brk;

    if (basepage*256 + CPM_LENGTH + 6*N_VEC >= dptable) {
	fprintf(stderr, "Not enough space in simulated ram to accommodate base at %02lX\n",
		basepage);
	exit(1);
    }

    for (i = 0; i < N_VEC; i++) { /* build bios jump table */
	BYTE *p = ram + bios_base + 3*i;
	p[0] = 0xc3;		/* jump */
	p[1] = 3*(i+N_VEC);
	p[2] = (bios_base >> 8);
	p[3*N_VEC] = 0x76;	/* halt */
	p[3*N_VEC+1] = i;	/* function code */
	p[3*N_VEC+2] = 0xc9;	/* return */
    }
    bios_top = bios_base + 6*N_VEC;
    i = (dptable - bios_top + 7) / 8;
    global_alv = xmalloc(i);
    memclr(global_alv, i);
}
#endif

/* getopt declarations */
extern int getopt();
extern char *optarg;
extern int optind, opterr, optopt;

int
main(int argc, char **argv)
{
    int c, gostart;
#ifdef BIOS
    int goopt;
#endif

    progname = argv[0];

    puts("\nYet Another Z80 Emulator version " VERSION
	 ", Copyright 1995,1998 Frank D. Cringle.\n"
	 "yaze comes with ABSOLUTELY NO WARRANTY; for details\n"
	 "see the file \"COPYING\" in the distribution directory.\n");

    while ((c = getopt(argc, argv, "b:l:s:"
#ifdef BIOS
                                           "p:vz:"
#endif
                                                  )) != EOF)
	switch (c) {
	case 'b':
	    bootfile = optarg;
	    break;
	case 'l':
	    loadadr = strtol(optarg, NULL, 16);
	    break;
	case 's':
	    startup = optarg;
	    break;
#ifdef BIOS
	case 'p':
	    basepage = strtol(optarg, NULL, 16);
	    break;
	case 'v':
	    vflag = 1;
	    break;
	case 'z':
	    z3env = strtol(optarg, NULL, 16);
	    break;
#endif
	case '?':
	    usage();
	}

#ifdef MMU
    for (c=0; c<16; ++c) pagetable[c]=ram+(c<<12);
#endif
#ifdef BIOS
    ccp_base = basepage * 256;
    bdos_base = ccp_base + CCP_LENGTH;
    bios_base = bdos_base + BDOS_LENGTH;
    if (z3env == bios_base)
	bios_base += 1024;
#endif

    load_cpm(bootfile);
#ifdef BIOS
    setvectors();

    if (vflag) {
	printf("bootfile:	%s\n", bootfile);
	if (loadadr >= 0)
	    printf("loadadr:	%4lx\n", loadadr);
	printf("startup file:	%s\n"
	       "basepage:	%2lx\n"
	       "ccp_base:	%4hx\n"
	       "bdos_base:	%4hx\n"
	       "bios_base:	%4hx\n"
	       "bios_top:	%4hx\n",
	       startup, basepage, ccp_base,
	       bdos_base, bios_base, bios_top);
	if (z3env)
	    printf("z3env:		%4lx\n", z3env);
	printf("dptable:	%4hx\n"
	       "dirbuf:		%4hx\n",
	       dptable, dirbuff);
    }
#endif

    gostart = bios_init(startup);

#ifdef BIOS
    for (goopt = 0; !goopt && optind < argc; optind++)
	goopt = docmd(argv[optind]);

    if (!gostart && !goopt)
	monitor(0);
#endif

#ifdef DEBUG
    do {
	FASTWORK adr = simz80(pc);
	if (adr & 0x10000)
	    monitor(adr);
	else
	    bios(ram[adr]);
    } while (1);
#elif !BIOS
    simz80(pc);
    fprintf(stderr,"HALT\n\r");
    fprintf(stderr,"PC   SP   IR   IX   IY   AF   BC   DE   HL   AF'  BC'  DE'  HL'\n\r");
    fprintf(stderr,"%04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x\n\r",pc,sp,ir,ix,iy,af[af_sel],regs[regs_sel].bc,regs[regs_sel].de,regs[regs_sel].hl,af[1-af_sel],regs[1-regs_sel].bc,regs[1-regs_sel].de,regs[1-regs_sel].hl);
    for (c=0; c<16; ++c) fprintf(stderr,"F%-4x",c);
    fprintf(stderr,"\r\n");
    for (c=0; c<16; ++c) fprintf(stderr,"%s%04x",c?" ":"",(pagetable[c]-ram)>>12);
    fprintf(stderr,"\r\n");
    exit(0);
#else
    do bios(ram[simz80(pc)]); while (1);
#endif
}

#ifndef USE_GNU_READLINE
void *
xmalloc(size_t size)
{
    void *p = malloc(size);

    if (p == NULL) {
	fputs("insufficient memory\n", stderr);
	exit(1);
    }
    return p;
}
#endif
